src/components/inputs/money-input/money-input.spec.js

```diff
@@ -345,15 +341,24 @@ describe('MoneyInput', () => {
     expect(getByTestId('money-input-container')).toHaveTextContent('EUR');
   });
 
-  it('should allow changing the amount', () => {
-    const onChange = jest.fn();
+  it('should allow changing the amount with numbers', () => {
+    let event;
+
+    const onChange = e => {
```

Comment:
courtesy of @pa3

Reply:
Just to shed some light on that: once creation of `fakeEvent` was removed from `handleAmountChange` the test started to check agains actual React Synthetic Events. By the time of expectation is executed Synthetic Event is already returned back to pool and all it fields is nullified, which leads to failing tests and lot's of warnings in console about us messing with Synthetic Events. 

That is just work-around, and maybe we'll be able to come with some utility function to deal with this kind of issue.

Weird stuff, that in theory, test should start passing once we add the body to jest.fn, like this:

```js
  it('should allow changing the amount', () => {
    const onChange = jest.fn(e => e.persist()); /// <====== here
    const { getByLabelText } = render(<TestComponent onChange={onChange} />);

    const event = { target: { value: '12' } };
    fireEvent.change(getByLabelText('Amount'), event);

    // onChange should be called with the updated amount
    expect(onChange).toHaveBeenCalledWith({
      persist: expect.any(Function),
      target: {
        name: 'some-name.amount',
        value: '12',
      },
    });
  });
```

But once I do this, jest starts to hang on trying to execute `toHaveBeenCalledWith` which looks crazy and I have no idea why is that happening.

Reply:
I got curious and looked into this a bit.

I dumbed the problem down and the following code is working for me:

```jsx
  it('should allow changing the amount with numbers', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(
      <TestComponent
        onChange={onChange}
        value={{
          currencyCode: 'EUR',
          amount: '13.00',
        }}
      />
    );

    fireEvent.change(getByLabelText('Amount'), { target: { value: '12' } });
    expect(onChange).toHaveBeenCalledTimes(1);
  });
```

This means the hanging is not because of the event itself. If we change the test to

```diff
-     const onChange = jest.fn();
+    const onChange = jest.fn(event => {
+      event.persist();
+    });
```

then it still works.

But it hangs again if we now change

```diff
-    expect(onChange).toHaveBeenCalledTimes(1);
+    expect(onChange).toHaveBeenCalledWith({ foo: true });
```

My assumption is that this happens because there are cycling references inside the event and so the `toHaveBeenCalledWith` follows this loop infinitely.

If we do this instead, the test will pass (and not hang)

```diff
-    expect(onChange).toHaveBeenCalledTimes(1);
+    expect(onChange.mock.calls[0][0].target.name).toBe('some-name.amount');
```

This would also work:

```diff
-    expect(onChange).toHaveBeenCalledTimes(1);
+    expect(onChange.mock.calls[0][0].target).toEqual(
+      expect.objectContaining({ name: 'some-name.amount' })
+    );
```

This brings us to the next problem. We can not verify the value with this approach. If we try to read `onChange.mock.calls[0][0].target.value`, then we will see the old value. This is because `MoneyInput` is a controlled component and the current test never passes the original value back in as a prop. So the amount input (what we see as `event.target`) will always keep the same value - even when `onChange` is called correctly, because our test is not passing the value back in - and because `event.target` is a reference to the DOM node itself - whose value never updated.

So.. what @pa3 came up with seems to be a good workaround, even if slightly awkward. Testing plain APIs of components doesn't seem to be a strength of `react-testing-library`, or maybe we just got something totally backwards.